lxd/metrics.go

```diff
@@ -0,0 +1,146 @@
+package main
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/metrics"
+	"github.com/lxc/lxd/lxd/response"
+	log "github.com/lxc/lxd/shared/log15"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+var metricsCmd = APIEndpoint{
+	Path: "metrics",
+
+	Get: APIEndpointAction{Handler: metricsGet, AccessHandler: allowProjectPermission("containers", "view")},
+}
+
+var metricsGetMu sync.Mutex
+var metricsGetBuilding bool
+
+// swagger:operation GET /1.0/metrics metrics metrics_get
+//
+// Get metrics
+//
+// Gets metrics of instances.
+//
+// ---
+// produces:
+//   - text/plain
+// parameters:
+//   - in: query
+//     name: project
+//     description: Project name
+//     type: string
+//     example: default
+// responses:
+//   "200":
+//     description: Metrics
+//     schema:
+//       type: string
+//       description: Instance metrics
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func metricsGet(d *Daemon, r *http.Request) response.Response {
+	// Prevent concurrent comparison/modification of metricsLastFetchedTime and access of d.metrics.
+	metricsGetMu.Lock()
```

Comment:
Oh, I see, you have a lock here as a global mutex. I think it'd be cleaner for that mutex to be closer to the locked data, so as part of the daemon struct.